breach-logos.js

```diff
@@ -0,0 +1,86 @@
+"use strict";
+
+const got = require("got");
+const fs = require("fs");
+const { promisify } = require("util");
+const writeFile = promisify(fs.writeFile);
+
+/* eslint-disable node/no-unpublished-require */
+const imagemin = require("imagemin");
+const pngQuant = require("imagemin-pngquant");
+/* eslint-enable node/no-unpublished-require */
+
+const mozlog = require("./log");
+const log = mozlog("breach-logos");
+
+
+const BreachLogos = {
+
+  remoteDir: "https://haveibeenpwned.com/Content/Images/PwnedLogos/",
+  localDir: "./public/img/logos/",
+
+  async minifyLogoData(logoData) {
+    return await imagemin.buffer(logoData, {
+      plugins: [pngQuant({quality: 10})],
+    });
+  },
+
+  async downloadLogo(fileName) {
+    const remoteSrc = `${this.remoteDir}${fileName}`;
+    const localSrc = `${this.localDir}${fileName}`;
+    let fileIsBig = false;
+    let logoData = [];
+
+    const stream = await got.stream(remoteSrc, {
+      retries: 3,
+      timeout: {
+        connect: 1000,
+        socket: 1000,
+        request: 5000,
+      },
+    });
+
+    stream.on("data", incomingData => logoData.push(incomingData));
+    stream.on("response", res => {
+
+        // Check response and error out on certain conditions.
+        if (res.headers["content-type"] !== "image/png") { /* eslint-disable quotes */
+          throw new Error(`response.headers["content-type"] !== "image/png"`); /* eslint-enable quotes */
+        }
+        if (res.statusCode !== 200) {
+          throw new Error(`Status: ${res.statusCode}`);
+        }
+        if (parseInt(res.headers["content-length"]) > 1000000) {
+          throw new Error(`${fileName} is too large. Downloaded Aborted.`);
+        }
+        if (parseInt(res.headers["content-length"]) > 10000) {
```

Comment:
Not sure if we want constants for the magic numbers here (and L53 above). I'm not sold either way, but this is basically just checking if the file size is around 9.75 KB, and if so, we should optimize the image.
Not sure if we should just always optimize the image, regardless of file size, or if that is too aggressive.

Reply:
I'll try this again with imagemin-pngquant and see how it goes with optimizing all images by default no matter their size. I encountered (possibly with imagemin-pngquant, possibly with one of the other set-ups I tried) two strange things when optimizing images that were already quite small: in some cases they turned into black squares, and in other cases they actually increased in file size.